src/metabase/upload.clj

```diff
@@ -513,6 +514,16 @@
            (field->db-type driver field->new-type)
            args)))
 
+(defn- invalidate-cached-models! [table]
+  ;; Find all the models for which this is the "primary table", and there is no meaningful filtering or transformation.
+  (let [model-ids (->> (t2/select [:model/Card :id :dataset_query] :table_id (:id table) :type :model)
+                       (remove (fn [{query :dataset_query}]
+                                 ;; Notably there are no joins or custom columns, which might be expensive.
+                                 (every? #{:source-table :fields :limit} (keys query))))
```

Comment:
1. This line depends on the internals of `:dataset_query`, which we should try to limit to MLv2 these days. See `no-joins` in this query for an example.
2. This feels like it will miss models that we'd still want to invalidate, e.g. a simple filter might be applied to the CSV table that excludes certain dirty rows or something. Also a model that just aggregates a CSV upload table should probably be invalidated too. I feel like we can be on the safe side and invalidate more models than we need to. If someone updates a table outside metabase, sure, it's expected that metabase shouldn't update models. But if someone updates a table inside metabase, we should expect that they want to see that data immediately.
3. This will return a false positive match for a model that wraps a more complex model. e.g. consider creating a model using the Orders table joined with People. Then make another model with a simple query selecting from that model, only applying a limit. It will match `model-ids` above, even though the actual query being executed is more complex than a simple select. The solution is to use MLv2 to get `lib/source-table-id` as is used here: https://github.com/metabase/metabase/blob/3c71e3a473a440e17c5fb0e90e35c5ab2e55aeee/src/metabase/upload.clj#L727

My gut feeling now is that simpler is better and we actually want to invalidate the model cache for any model that uses CSV upload data, regardless of the complexity of the query. Using just `table_id` to invalidate all models that use the upload table seems like a good starting point. Ideally we'd want to update even models that use the uploaded table in joins and in native queries too, but that's harder at the moment. Probably @luizarakaki has an opinion on this too